Fixed MagicalRecordModelService (mostly); UI improvements

WERUreo opened the pull request:
The MagicalRecordModelService is mostly working.  Working with the Tour entity is done.  That includes creating, deleting, and finding tours.  The user can find all the tours or a specific tour given that tour's UUID.  The user can also add and load HistoricLocation entities.  Deleting locations still needs to be looked at.

While going through this, I made the connections between the Dashboard/Add Tour views, the LocationList view, and the CurrentTour view.  On the CurrentTour view, the user can view all locations that have been added to that tour both in list and map form.  The user can also tap the Add bar button to go back to LocationList and add more locations to that tour.

Finally, I made some UI improvements.  I modified the layout of the LocationList view to look more native.  I added a custom font (Ubuntu) to use throughout the app, if we decide to keep it.

WERUreo committed:
Fixed working with Tours in MagicalRecordModelService

I've started to go through the MagicalRecordModelService to get it working correctly.  First stop was working with Tours.  I've made the appropriate changes to create tours with an NSUUID, a new property added to the Tour entity.  I've also added functions to find tours by UUID as well as delete tours.

While I was in there, I modified the data services to not use a singleton.

WERUreo committed:
More work on MagicalRecordService; changes to LocationList

I have been able to get most of the MagicalRecordModelService working.  There are still a couple of functions that need to be addressed.  I have also modified the layout of the LocationList view to give it a more native feel.  I removed the placeholder image that was there and am now using my dynamically loaded random placeholder image.  I need to go back and put the "Closest to Me" sort option back in; I haven't figured out how to display that yet.  I'm thinking it will be another bar button item.

I've added a custom font (Ubuntu) and have started to use that font throughout the app.  This is mainly for my appeasement as I don't really like the system font.  We can change it later if others don't like it.

On the CurrentTour view, I've implemented displaying the tours that have been saved for that tour both in the table view and map view.  I've also added an Add Location bar button which brings the user back to the LocationList view to add more locations.

LondonAtlas commented on Orlando Walking Tours/Controllers/AddTourVC.swift:
```diff
@@ -24,6 +23,7 @@ class AddTourVC: UIViewController, UITextFieldDelegate
 
     let notificationCenter = NSNotificationCenter.defaultCenter()
     var uuid: NSUUID?
+    var modelService: ModelService!
```
either do what you did with Notification Center and initialise inline or change the ! to ?, because things could explode.

LondonAtlas commented on Orlando Walking Tours/Controllers/AddTourVC.swift:
```diff
@@ -68,14 +70,8 @@ class AddTourVC: UIViewController, UITextFieldDelegate
-            tour?.uuid = self.uuid?.UUIDString
-        })
+        self.uuid = NSUUID()
+        modelService.createTour(uuid: self.uuid!, title: tourNameTextField.text!)
```
safely unwrap tourNameTextField.text!

WERUreo replied:
I'm thinking about modifying the `createTour()` function a little...  i don't quite like having to create a UUID first and then passing it into the create function.  I think i might move the creating into the function itself and just put the UUID into the completion handler instead.  I feel like that would be better encapsulation.  Thoughts?

LondonAtlas commented on Orlando Walking Tours/Controllers/AddTourVC.swift:
```diff
@@ -127,8 +123,10 @@ class AddTourVC: UIViewController, UITextFieldDelegate
             {
-                let tour = Tour.MR_findFirstByAttribute("uuid", withValue: (self.uuid?.UUIDString)!)
-                vc.tour = tour
+                modelService.findTour(byUUID: self.uuid!)
```
safely unwrap self.uuid

LondonAtlas commented on Orlando Walking Tours/Controllers/CurrentTourVC.swift:
```diff
@@ -23,6 +23,11 @@ class CurrentTourVC: UIViewController
 
     var tour: Tour?
+    var locations = [HistoricLocation]()
+    var modelService: ModelService!
```
same as above

LondonAtlas commented on Orlando Walking Tours/Controllers/CurrentTourVC.swift:
```diff
@@ -32,6 +37,102 @@ class CurrentTourVC: UIViewController
+    {
+        let location = self.locations[indexPath.row]
+
+        let cell = tableView.dequeueReusableCellWithIdentifier(CurrentLocationTableViewCell.reuseIdentifier, forIndexPath: indexPath) as! CurrentLocationTableViewCell
```
don't use as! safely unwrap with as?

LondonAtlas commented on Orlando Walking Tours/Controllers/DashboardVC.swift:
```diff
@@ -24,6 +23,7 @@ class DashboardVC: UIViewController, UICollectionViewDataSource, UICollectionVie
 
     var tours = [Tour]()
     var isDeletionModeActive = false
+    var modelService: ModelService!
```
same as above

LondonAtlas commented on Orlando Walking Tours/Controllers/LocationListVC.swift:
```diff
@@ -54,7 +55,7 @@ class LocationListVC: UIViewController
     ////////////////////////////////////////////////////////////
 
-    var viewMode : ViewMode!
+    var viewMode : ViewMode! = .List
```
you don't  need the ! since it is being initialised instantly.

LondonAtlas commented on Orlando Walking Tours/Controllers/LocationListVC.swift:
```diff
@@ -66,6 +67,7 @@ class LocationListVC: UIViewController
     var tourObjectID: NSManagedObjectID?
     var locations = [HistoricLocation]()
     var modelService: ModelService!
+    var dataService: DataService!
```
same as above

LondonAtlas commented on Orlando Walking Tours/Controllers/LocationListVC.swift:
```diff
@@ -80,23 +82,19 @@ class LocationListVC: UIViewController
-            self.sortClosestToMe()
-            self.tableView.reloadData()
-        }
+//        closestToMe.didTouch =
```
delete unused code. keep it clean.

WERUreo replied:
Yeah, I left this stuff in because I need to re-implement it.  When I changed the layout, I completely remove the TouchableLabel that handled this functionality.  I will add a TODO: to remind me to work on that.

LondonAtlas commented on Orlando Walking Tours/Controllers/LocationListVC.swift:
```diff
@@ -120,6 +118,19 @@ class LocationListVC: UIViewController
+    override func prepareForSegue(segue: UIStoryboardSegue, sender: AnyObject?)
+    {
+        if let navController = segue.destinationViewController as? UINavigationController,
+           let vc = navController.topViewController as? CurrentTourVC where segue.identifier == "ShowCurrentTourSegue"
```
as? 👌

WERUreo replied:
BTW, I picked up this style from your code in the Trees app. :) :+1:

LondonAtlas replied:
I missed this in the first look over but there is no call to super.prepareForSegue()

LondonAtlas commented on Orlando Walking Tours/Controllers/LocationListVC.swift:
```diff
@@ -260,20 +280,28 @@ extension LocationListVC : UITableViewDataSource, UITableViewDelegate
         
         let location = locations[indexPath.row]
-        let locationCell = tableView.dequeueReusableCellWithIdentifier(String(LocationTableViewCell)) as! LocationTableViewCell
+        let locationCell = tableView.dequeueReusableCellWithIdentifier(LocationTableViewCell.reuseIdentifier) as! LocationTableViewCell
```
as!

LondonAtlas commented on Orlando Walking Tours/Controllers/LocationListVC.swift:
```diff
@@ -284,10 +312,10 @@ extension LocationListVC : UITableViewDataSource, UITableViewDelegate
-        locationDetailVC.modalTransitionStyle = .FlipHorizontal
-        locationDetailVC.location = self.locations[indexPath.row]
-        self.navigationController?.pushViewController(locationDetailVC, animated: true)
+//        let storyboard = UIStoryboard(name: "LocationDetails", bundle: nil)
```
delete unused code

LondonAtlas commented on Orlando Walking Tours/Services/InMemoryModelService.swift:
```diff
@@ -28,16 +28,35 @@ class InMemoryModelService : ModelService {
     }
+
+    func deleteTour(tour tour: Tour, completion: ModelServiceCompletionHandler?) {
+        for localTour in tours {
```
instead of looping over all tours maybe do: tours.filter { $0.uuid == tour.uuid }

then run a loop over and delete all returned objects (probably just the one), it will be faster.

WERUreo replied:
You know, I'm sort of thinking about just removing this entire file.  This was originally put in place because the model service that dealt with the actual persistent store wasn't ready.  Since I've been working on getting that service to work, this service never gets called at all.

LondonAtlas commented on Orlando Walking Tours/Views/LocationTableViewCell.swift:
```diff
@@ -7,11 +7,62 @@
+    // MARK: - Properties
+    ////////////////////////////////////////////////////////////
+
+    var locationId: String!
```
same as above

WERUreo replied:
I need to find out from @gregsbarr what this locationId is used for.  It may be something that can be removed completely.

LondonAtlas commented on Orlando Walking Tours/Views/LocationTableViewCell.swift:
```diff
@@ -7,11 +7,62 @@
+
+    ////////////////////////////////////////////////////////////
+
+    func populateCell(image: UIImage)
```
all UIImageViews accept UIImage? , add a ? to the end of the passed variable class.

LondonAtlas commented on Orlando Walking Tours/Views/LocationTableViewCell.swift:
```diff
@@ -7,11 +7,62 @@
+        // TODO: Image view for cell should be random photo of a location from the tour
+        request = UIImage.getPlaceholderImage(sized: Int(width), by: Int(height))
+        { image in
+            self.populateCell(image!)
```
don't force unwrap an image, many bad times ahead.

WERUreo committed:
Fixed some code review defects

Replaced some unsafe force unwrapped optionals, refactored a function in the MagicalRecordModelService

LondonAtlas commented on Orlando Walking Tours/Controllers/DashboardVC.swift:
```diff
@@ -197,8 +194,6 @@ class DashboardVC: UIViewController, UICollectionViewDataSource, UICollectionVie
     {
-        print(segue.identifier)
-        print(segue.destinationViewController)
         if let navController = segue.destinationViewController as? UINavigationController,
```
There is no call to super here. probably wise to call super.prepareForSegue()

LondonAtlas commented on Orlando Walking Tours/Controllers/CurrentTourVC.swift:
```diff
@@ -32,6 +37,104 @@ class CurrentTourVC: UIViewController
+
+    override func prepareForSegue(segue: UIStoryboardSegue, sender: AnyObject?)
+    {
+        if let navController = segue.destinationViewController as? UINavigationController,
```
No call to super.prepareForSegue()

LondonAtlas commented on Orlando Walking Tours/Controllers/CurrentTourVC.swift:
```diff
@@ -32,6 +37,104 @@ class CurrentTourVC: UIViewController
+
+    override func viewDidAppear(animated: Bool)
+    {
+        self.loadLocations()
```
call super.viewDidAppear()

WERUreo committed:
More defect fixes

Add calls to appropriate superclass methods where needed (prepareForSegue(), viewDidAppear())

WERUreo committed:
Without these changes, the app crashes

WERUreo committed:
Adding one last TODO
